src/Domain/BeginnerInterface.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace PhpTuf\ComposerStager\Domain;
+
+/**
+ * Begins the staging process by copying the active directory to the staging directory.
+ */
+interface BeginnerInterface
+{
+    /**
+     * Begins the staging process.
+     *
+     * @param string $activeDir
+     *   The path to the active directory, absolute or relative to the working
+     *   directory (CWD), e.g., "/var/www/public" or "public".
+     * @param string $stagingDir
+     *   The path to the staging directory, absolute or relative to the working
+     *   directory (CWD), e.g., "/var/www/staging" or "staging".
+     * @param callable|null $callback
+     *   An optional PHP callback to run whenever there is some output available
+     *   on STDOUT or STDERR. Example:
+     *
+     *   ```php
+     *   $callback = function (string $type, string $buffer) {
+     *       if ($type === \PhpTuf\ComposerStager\Domain\OutputType::ERR) {
+     *           echo 'ERR > ' . $buffer;
+     *       }
+     *       if ($type === \PhpTuf\ComposerStager\Domain\OutputType::OUT) {
+     *           echo 'OUT > ' . $buffer;
+     *       }
+     *   }
+     *   ```
+     *
+     * @throws \PhpTuf\ComposerStager\Exception\DirectoryAlreadyExistsException
+     *   If the staging directory already exists.
+     * @throws \PhpTuf\ComposerStager\Exception\DirectoryNotFoundException
+     *   If the staging directory is not found.
+     * @throws \PhpTuf\ComposerStager\Exception\ProcessFailedException
+     *   If the command process doesn't terminate successfully.
+     */
+    public function begin(string $activeDir, string $stagingDir, ?callable $callback = null): void;
+
+    /**
+     * Determines whether or not the given active directory exists.
+     *
+     * @param string $activeDir
+     *   The path to the active directory, absolute or relative to the working
+     *   directory (CWD), e.g., "/var/www/public" or "public".
+     *
+     * @return bool
+     */
+    public function activeDirectoryExists(string $activeDir): bool;
```

Comment:
I'm wondering if we really need `activeDirectoryExists()` and `stagingDirectoryExists()` to be public interface methods. Don't those pretty much assume a filesystem-based implementation, a bit outside of the scope of the interface?

Reply:
Yes, they probably assume a filesystem-based implementation, but without an alternate implementation to compare I don't know enough to be more flexible or abstract. But I think the methods are important to the public interface so that clients, e.g., Drupal, can do readiness checks without committing to performing the operation. For instance, Drupal might want to display configuration errors and hide the process button or whatever. I considered some kind of `::assertPreconditions()` that throws exceptions, but they're not really exceptional conditions; and passing back some kind of array of failed conditions seems ugly. I could add a comment explaining the reason for the methods if you think it's confusing.

Reply:
One idea here might be to do something like this:

```
public function isActiveAreaReady($...implementation_specific_arguments): void;

public function isStagingAreaReady($...implementation_specific_arguments): void;
```

Then we're not making filesystem assumptions, but we ARE codifying the idea that there are distinct "active" and "staging" areas which can be in different states of readiness.

Reply:
I like having a single check for general readiness, but I think clients will still need to know _what_ isn't ready, e.g., is the staging directory absent or not writable? If we use an `isReady` pattern, we need to supplement it with something to get those details.

But I'm also not thrilled about the client having to know what constitutes readiness, i.e., existence and writability of the given directory.

We could create some kind of `::checkReadiness()` that returns an array of some kind of `ReadinessError` objects. I like the consistency of interface and encapsulation of implementation details. (This approach would be much more susceptible to swapping out directory/area implementations.) But a client like Drupal probably wants to display a table of readiness conditions with a pass/fail status for each. That means it would want to iterate over all preconditions or even ask for each specifically. We could definitely provide such a facility, but I don't know if we want to delay this PR that long.

Maybe for now we should just remove these two methods from the public interface so nobody starts depending on them. After all, now that I think of it, there's currently no code using these methods now that we changed the console command to just try the operation and catch exceptions. Then we can come back later and augment the interface when we know more about how it might be used. @effulgentsia, would you like to weigh in here?

Reply:
Looping in @tedbow because we have been working on something like what you describe, for core (in https://git.drupalcode.org/project/drupal/-/merge_requests/66). So your ideas might indeed dovetail beautifully.

Reply:
> After all, now that I think of it, there's currently no code using these methods now that we changed the console command to just try the operation and catch exceptions.

I like this idea. Since we have `\PhpTuf\ComposerStager\Exception\PathException::getPath()` and the staging and active directories are required options from these commands the caller would be able to figure out if it is staging or active directory that has the problem.

Re https://git.drupalcode.org/project/drupal/-/merge_requests/66 I am not sure we want to co-mingle the command here with the needs of Drupal  core. Drupal core has the problem where it needs to periodically check if the system is ready to do an update with actually wanting to attempt one. 

I think the duplication of having Drupal have to check in it's readiness checks if the staging directory it intends to use already exists is not a problem.  Or Drupal could just adaptive where if the directory it would normally for staging already exists create an alternate directory or use based on timestamp. In either case I don't think Drupal should have lean on this library for that kind of readiness check and I don't think this library needs the API concept of readiness checks just because Drupal will have them.

Reply:
Per offline discussion: We will remove these methods for now and think about the best way to handle preconditions later.